src/sentry/api/endpoints/organization_transaction_anomaly_detection.py

```diff
@@ -0,0 +1,137 @@
+        query_start: datetime representing start of query window
+        query_end: datetime representing end of query window
+        granularity: granularity to use (in seconds)
+    """
+    anomaly_detection_range = end - start
+
+    if anomaly_detection_range > timedelta(days=14):
+        snuba_range = timedelta(days=90)
+        granularity = 3600
+
+    elif anomaly_detection_range > timedelta(days=1):
+        granularity = 1200
+        snuba_range = timedelta(days=28)
+
+    else:
+        snuba_range = timedelta(days=14)
+        granularity = 600
+
+    additional_time_needed = snuba_range - anomaly_detection_range
+    now = datetime.utcnow().astimezone(timezone.utc)
+    start_limit = now - timedelta(days=90)
+    end_limit = now
+    start = max(start, start_limit)
+    end = min(end, end_limit)
+    # By default, expand windows equally in both directions
+    window_increase = additional_time_needed / 2
+    query_start, query_end = None, None
+
+    # If window will go back farther than 90 days, use today - 90 as start
+    if start - window_increase < start_limit:
+        query_start = now - timedelta(days=90)
+        additional_time_needed -= start - query_start
+        window_increase = additional_time_needed
+    # If window extends beyond today, use today as end
+    if end + window_increase > end_limit:
+        query_end = now
+        additional_time_needed -= query_end - end
+        window_increase = additional_time_needed
+
+    query_start = query_start or max(start - window_increase, start_limit)
+    query_end = query_end or min(end + window_increase, end_limit)
+
+    return MappedParams(
+        query_start,
+        query_end,
+        granularity,
+    )
+
+
+class OrganizationTransactionAnomalyDetectionEndpoint(OrganizationEventsEndpointBase):
+    def has_feature(self, organization, request):
+        return features.has(
+            "organizations:performance-anomaly-detection-ui", organization, actor=request.user
+        )
+
+    def get(self, request: Request, organization) -> Response:
+        if not self.has_feature(organization, request):
+            return Response(status=404)
+
+        start, end = get_date_range_from_params(request.GET)
+        time_params = get_time_params(start, end)
+        query_params = self.get_snuba_params(request, organization)
+        query = request.GET.get("query")
+        query = f"{query} event.type:transaction" if query else "event.type:transaction"
+
+        datetime_format = "%Y-%m-%d %H:%M:%S"
+        ads_request = {
+            "query": query,
+            "params": query_params,
+            "start": start.strftime(datetime_format),
+            "end": end.strftime(datetime_format),
+            "granularity": time_params.granularity,
+        }
+
+        # overwrite relevant time params
+        query_params["start"] = time_params.query_start
+        query_params["end"] = time_params.query_end
+
+        snuba_response = timeseries_query(
```

Comment:
`timeseries_query()` could raise a `ValidationError` exception. You'll want to wrap this with try/catch and return a 400 status code error.

Reply:
1. Imported `ValidationError` from `django.core.validators`. Please advise if otherwise.
2. Also handled `InvalidSearchQuery`. Hope that's alright.

Reply:
@wmak can you confirm if this is fine?